### ISSUE TITLE
array or ndarray as argument type documentation?

### ISSUE THREAD

**Issue by rfezzani** [scikit-image/scikit-image#4407]
## Description

I found that argument type documentation is not consistent for numpy arrays over `skimage` functions docstings:

- `array` is used 198 times,
- `ndarray` 118 times,
- and `numpy array` 5 times.

Should we think about defining a convention for this?

Here are the cleaned output of `grep`:
[array.txt](https://github.com/scikit-image/scikit-image/files/4098948/array.txt)
[ndarray.txt](https://github.com/scikit-image/scikit-image/files/4098949/ndarray.txt)

**Comment by grlee77** [comment]
I see use both `array` and `ndarray` used with the docstrings of NumPy and SciPy themselves, although `ndarray` seems to be more common. I don't see a clear preference for either case specified in the [NumPy documenting guidelines](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)

The other case you didn't mention is `array_like` which should technically be used rather than `ndarray` when the function accepts other types of sequences as well (e.g. it may call `np.asarray` or `np.atleast1d`, etc. internally).

**Comment by rfezzani** [comment]
Thank you @grlee77 for linking to numpy doc guidelines. I did not considered `array_like` since it is used to describe arguments that can be converted to `ndarray` (i.e. scalar types, sequence types) according to these guidelines.
My preference goes to `ndarray`, since it is the name of the numpy array class while `array` is a function for numpy array creation.

**Comment by jni** [comment]
In my head, `ndarray` meant "a NumPy ndarray object", while `array` meant "array-like". Perhaps we should use that convention or similar? Specifically, I am thinking that any function that uses Cython needs an ndarray, but most that don't can still work with CuPy arrays or dask arrays. It would be great if we could document this distinction somehow.

**Comment by rfezzani** [comment]
I like your point of view @jni, but why not use the more conventional `array_like` instead of `array` then?
If we decide to go this way, we will have to call `np.asarray` whenever an argument is declared as `array_like`. This make me think of the `matrix` issue #4383.

**Comment by jni** [comment]
Well, it's actually a slightly different issue. We *don't* want to call `np.asarray` on a Dask array. We just want to call `np.sum`, `np.max`, `arr @ vals`, etc as usual, and let [`__array_function__`](https://numpy.org/neps/nep-0018-array-function-protocol.html) and related protocols do their magic.

This is annoyingly subtle! LOL

**Comment by rfezzani** [comment]
I confess that I am not aware of the position of skimage concerning dask :-S, I understand from your comment that we want to be compatible with such libraries.
Well, it is becoming a little bit more subtle in fact :-)
An option would be to use `dask.array.asarray` if dask is installed and `np.asarray` otherwise.

**Comment by jni** [comment]
No, we don't want to cater to dask specifically, but, where possible, we want to be able to chain skimage functions with dask arrays *or other duck arrays* transparently. The NEP-18 protocal was designed with this in mind. We don't actually have a policy on this at all just yet, I'm just being forward-looking.

We've had users who want to bring this future about, see e.g. #3823

**Comment by rfezzani** [comment]
Yes, I just realized my mistake after rereading my answer :-)
The option is then to call `np.asarray` only if the input is not a dask array :-p

**Comment by rfezzani** [comment]
I found that `iterable` is also used as an alternative to `array_like`:
```
skimage/filters/thresholding.py:846:    axis_sizes : iterable of int
skimage/filters/thresholding.py:871:    w : int, or iterable of int
skimage/filters/thresholding.py:935:    window_size : int, or iterable of int, optional
skimage/filters/thresholding.py:1000:    window_size : int, or iterable of int, optional
skimage/filters/ridges.py:89:    sigmas : iterable of floats
skimage/filters/ridges.py:183:    sigmas : iterable of floats, optional
skimage/filters/ridges.py:285:    sigmas : iterable of floats, optional
skimage/filters/ridges.py:376:    sigmas : iterable of floats, optional
skimage/filters/ridges.py:523:    sigmas : iterable of floats, optional
skimage/draw/draw.py:15:    shape :  iterable of ints
skimage/draw/draw.py:18:    center : iterable of floats
skimage/draw/draw.py:20:    radii : iterable of floats
skimage/draw/draw.py:29:    rows : iterable of ints
skimage/draw/draw.py:31:    cols : iterable of ints
skimage/graph/spath.py:24:    p : iterable of int
skimage/graph/mcp.py:15:    start : iterable
skimage/graph/mcp.py:17:    end : iterable
skimage/io/collection.py:25:    ic : an iterable of images
skimage/_shared/utils.py:92:    val : scalar or iterable of scalars
skimage/_shared/utils.py:168:    ndim : int or iterable of ints
skimage/registration/_optical_flow_utils.py:22:    shape : iterable
skimage/transform/_warps.py:294:    center : iterable of length 2
skimage/segmentation/random_walker_segmentation.py:295:    spacing : iterable of floats, optional
```

**Comment by rfezzani** [comment]
Well, I think that I found a good compromise, what about considering that `array` is for data buffers (as `np.array` or `dask.array`) and that `array_like` is for iterable?

**Comment by lagru** [comment]
I agree. To put it in more words:

- `array_like` -> Requires nested structure (e.g. iterable) that can be passed to `np.asarray`. To me this implies that the input will be likely converted into a NumPy array internally.
- `array` -> Requires object which supports (parts of) NumPy's array API (e.g. but not exclusively `dask.array`). Implies that the object does not need to be converted with `np.asarray`.
- `ndarray` -> Requires explicit NumPy array. Implies that the object does not need to be converted with `np.asarray`.

Requirements for the dimensions should be orthogonal like they are right now with a leading shape description in parenthesis.

**Comment by rfezzani** [comment]
Exactly, and thank you @lagru for adding the `ndarray` case which corresponds, for example, to arguments of Cython functions ;-)
